server/src/main/java/org/elasticsearch/index/codec/tsdb/es819/ES819TSDBDocValuesConsumer.java

```diff
@@ -138,89 +143,126 @@ private long[] writeField(FieldInfo field, TsdbDocValuesProducer valuesProducer,
-
-            final long valuesDataOffset = data.getFilePointer();
-            // Special case for maxOrd of 1, skip writing the blocks
-            if (maxOrd != 1) {
-                final long[] buffer = new long[ES819TSDBDocValuesFormat.NUMERIC_BLOCK_SIZE];
-                int bufferSize = 0;
-                final TSDBDocValuesEncoder encoder = new TSDBDocValuesEncoder(ES819TSDBDocValuesFormat.NUMERIC_BLOCK_SIZE);
-                values = valuesProducer.getSortedNumeric(field);
-                final int bitsPerOrd = maxOrd >= 0 ? PackedInts.bitsRequired(maxOrd - 1) : -1;
-                for (int doc = values.nextDoc(); doc != DocIdSetIterator.NO_MORE_DOCS; doc = values.nextDoc()) {
-                    final int count = values.docValueCount();
-                    for (int i = 0; i < count; ++i) {
-                        buffer[bufferSize++] = values.nextValue();
-                        if (bufferSize == ES819TSDBDocValuesFormat.NUMERIC_BLOCK_SIZE) {
-                            indexWriter.add(data.getFilePointer() - valuesDataOffset);
-                            if (maxOrd >= 0) {
-                                encoder.encodeOrdinals(buffer, data, bitsPerOrd);
-                            } else {
-                                encoder.encode(buffer, data);
+        IndexOutput disiTempOutput = null;
+        String skipListTempFileName = null;
+        IndexedDISIBuilder docIdSetBuilder = null;
+        try {
+            if (numValues > 0) {
+                assert numDocsWithValue > 0;
+                // Special case for maxOrd of 1, signal -1 that no blocks will be written
+                meta.writeInt(maxOrd != 1 ? ES819TSDBDocValuesFormat.DIRECT_MONOTONIC_BLOCK_SHIFT : -1);
+                final ByteBuffersDataOutput indexOut = new ByteBuffersDataOutput();
+                final DirectMonotonicWriter indexWriter = DirectMonotonicWriter.getInstance(
+                    meta,
+                    new ByteBuffersIndexOutput(indexOut, "temp-dv-index", "temp-dv-index"),
+                    1L + ((numValues - 1) >>> ES819TSDBDocValuesFormat.NUMERIC_BLOCK_SHIFT),
+                    ES819TSDBDocValuesFormat.DIRECT_MONOTONIC_BLOCK_SHIFT
+                );
+
+                final long valuesDataOffset = data.getFilePointer();
+                // Special case for maxOrd of 1, skip writing the blocks
+                if (maxOrd != 1) {
+                    final long[] buffer = new long[ES819TSDBDocValuesFormat.NUMERIC_BLOCK_SIZE];
+                    int bufferSize = 0;
+                    final TSDBDocValuesEncoder encoder = new TSDBDocValuesEncoder(ES819TSDBDocValuesFormat.NUMERIC_BLOCK_SIZE);
+                    values = valuesProducer.getSortedNumeric(field);
+                    final int bitsPerOrd = maxOrd >= 0 ? PackedInts.bitsRequired(maxOrd - 1) : -1;
+                    if (enableOptimizedMerge && numDocsWithValue < maxDoc) {
+                        // TODO: which IOContext should be used here?
```

Comment:
I think we should use MERGE for this IOContext?

Reply:
I don't see a `MERGE` constant for io context. But I did the following instead: 45308e43a81dbd8775a1c95649cc927c1829ca81

I think this way we always get most appropriate io context? In case of merge it will be an io context for merging?